### PR TITLE
Device sync improvements

### PR DESCRIPTION
**Pull request by Skons**
This PR improves the way devices are synced. It has got a reverse engineered device sync implementation from the original app. It does this by using the DEVICE_NAME_REPLY, grabbing the ID of the device and send it back with a bogus device state. This will trigger DEVICE_STATUS_UPDATE to be sent to ELRO Connects. With this, ELRO Connects will get all the devices and their statuses within ~40 seconds after launch.

The protocol documentation has also been updated.

### CODE REVIEW

**Comment by Skons** [comment]
This PR solves issue #18

**Comment by Skons** [comment]
@hildensia not to push you, but do you have time to review this pr?

File: elro/hub.py [hunk 1]
```diff
@@ -50,6 +51,12 @@ async def sender_task(self):
         await self.sync_scenes(0)
         await self.get_device_names()
 
+        logging.info("Waiting until all devices are retreived")
+        await trio.sleep(5)
+        if len(self.devices_for_sync) > 0: #sync devices when there are devices known by name
```
Comment (hildensia) [inline elro/hub.py:56]: PEP8: 2 spaces befor and one after `#`

File: elro/hub.py [hunk 2]
```diff
@@ -50,6 +51,12 @@ async def sender_task(self):
         await self.sync_scenes(0)
         await self.get_device_names()
 
+        logging.info("Waiting until all devices are retreived")
+        await trio.sleep(5)
```
Comment (hildensia) [inline elro/hub.py:55]: why 5s? Could we react to the actual event somehow?

File: elro/hub.py [hunk 3]
```diff
@@ -209,6 +215,16 @@ async def handle_command(self, data):
             d_id = int(answer[0:4], 16)
             name_val = get_string_from_ascii(answer[4:])
 
+            #Build a list with known device names by device id
```
Comment (hildensia) [inline elro/hub.py:218]: PEP8: space after `#`

File: elro/hub.py [hunk 4]
```diff
@@ -209,6 +215,16 @@ async def handle_command(self, data):
...
+            try:
+                dev = self.devices_for_sync[d_id]
+            except KeyError:
+                logging.info(f"Unknown name from device id '{d_id}'")
+                self.devices_for_sync[d_id] = "0464AA00" #Bogus device status
+                return
+            await trio.sleep(0)
+
+            #Set the device name from this reply
```
Comment (hildensia) [inline elro/hub.py:227]: PEP8: Space after `#`

**Review by hildensia — APPROVED** [review]

**Comment by hildensia** [comment]
Only minor things. Can you run the `pep8` tool on your code to lint it?

File: elro/hub.py (hunk 2 above)
Comment (Skons) [reply elro/hub.py:55]: The problem faced here is that it is not known how many devices will be returned. So we can wait for the first device name to return, but then we still don't know how much devices will show up. The best thing to do imo, is to rerun this sync some times more often, but it is not necessary to do this every 30 seconds. It takes some time to implement that. Is it ok that this becomes an issue and it is resolved later?

**Pull request merged**